src/mnesia_api.erl

```diff
@@ -0,0 +1,313 @@
+                                   [filename:absname(File), node(), Reason]),
+            case Reason of
+                table_does_not_exist ->
+                    {error, {table_does_not_exist, String}};
+                _ ->
+                    {error, {cannot_dump, String}}
+            end
+    end.
+
+%% @doc Mnesia database restore
+%% This function is called from ejabberd_ctl, ejabberd_web_admin and
+%% mod_configure/adhoc
+restore(Path) ->
+    mnesia:restore(Path, [{keep_tables, keep_tables()},
+                          {default_op, skip_tables}]).
+
+%% @doc This function return a list of tables that should be kept from a
+%% previous version backup.
+%% Obsolete tables or tables created by module who are no longer used are not
+%% restored and are ignored.
+-spec keep_tables() -> [atom()].
+keep_tables() ->
+    lists:flatten([acl, passwd, disco_publish, keep_modules_tables()]).
+
+%% @doc Returns the list of modules tables in use, according to the list of
+%% actually loaded modules
+-spec keep_modules_tables() -> [[atom()]]. % list of lists
+keep_modules_tables() ->
+    lists:map(fun(Module) -> module_tables(Module) end,
+              gen_mod:loaded_modules()).
+
+%% TODO: This mapping should probably be moved to a callback function in each module.
+%% @doc Mapping between modules and their tables
+-spec module_tables(_) -> [atom()].
+module_tables(mod_announce) -> [motd, motd_users];
+module_tables(mod_irc) -> [irc_custom];
+module_tables(mod_last) -> [last_activity];
+module_tables(mod_muc) -> [muc_room, muc_registered];
+module_tables(mod_offline) -> [offline_msg];
+module_tables(mod_privacy) -> [privacy];
+module_tables(mod_private) -> [private_storage];
+module_tables(mod_pubsub) -> [pubsub_node];
+module_tables(mod_roster) -> [roster];
+module_tables(mod_shared_roster) -> [sr_group, sr_user];
+module_tables(mod_vcard) -> [vcard, vcard_search];
+module_tables(_Other) -> [].
+
+-spec get_local_tables() -> [any()].
+get_local_tables() ->
+    Tabs1 = lists:delete(schema, mnesia:system_info(local_tables)),
+    Tabs = lists:filter(
+             fun(T) ->
+                     case mnesia:table_info(T, storage_type) of
+                         disc_copies -> true;
+                         disc_only_copies -> true;
+                         _ -> false
+                     end
+             end, Tabs1),
+    Tabs.
+
+-spec dump_to_textfile(any(), any(),
+                      {error, atom()} | {ok, pid() | {file_descriptor, atom() | tuple(), _}}
+                      ) -> ok | {error, atom()} | {file_error, atom()}.
+dump_to_textfile(yes, Tabs, {ok, F}) ->
+    try
+        Defs = lists:map(
+                 fun(T) -> {T, [{record_name, mnesia:table_info(T, record_name)},
+                                {attributes, mnesia:table_info(T, attributes)}]}
+                 end,
+                 Tabs),
+        io:format(F, "~p.~n", [{tables, Defs}]),
+        lists:foreach(fun(T) -> dump_tab(F, T) end, Tabs),
```

Comment:
I think it would be better to put only this part in the`try...catch`, because we might be catching other unexpected errors as well, hiding the real reason behind them.